### ISSUE TITLE
Add symbolic username redirects

### ISSUE THREAD

**Issue by ScottPeterJohnson** [ScottPeterJohnson/purelymail-issues#14]
(This issue was imported from Gitea)
**Scott** on February 28, 2019:
When routing mail, ignore everything in an email address after a "+" or "_" symbol.
This allows giving out "tagged" addresses, e.g. user+string@gmail.com. Gmail supports +, but we'll add _ because the + trick is fairly well known.

This will also require enhancing custom domain routing to support this.

**Comment by ScottPeterJohnson** [comment]
Comment by **Scott** on April 9, 2019:
https://news.purelymail.com/posts/2019-04-08-sieve.html

Now implemented! Applies to all shared addresses.

We still need to add a good GUI for adding this to custom domains, but you can either:

* Use a catchall forward rule (if you only have one user on your domain)
* Use a specific prefix rule per-user (matching say, "user_" with prefix enabled)
* Add the following account-wide Sieve script:

```
require "subaddress";
require "variables";

if envelope "to" :user :matches "*" {
    set "user" "${1}";
}
if envelope "to" :domain :matches "*" {
    set "domain" "${1}";
}
redirect "${user}@${domain}";
```

Filters should now be able to match on the user/detail parts. Roundcube doesn't make it  very obvious you can do this though (there's a tiny disclosure triangle to the left of a condition).

**Comment by ScottPeterJohnson** [comment]
Comment by **jeremy** on May 21, 2019:
@Scott, I've added that Sieve script, but it has not taken immediate effect. I have 2 requests:

1. Add that Sieve script by default to all accounts that don't have Sieve scripts (so you don't mess up what people already have) so that custom domains will support this by default to match your domains.

2. Add a note on how long it takes for Sieve scripts to take effect on https://purelymail.com/manage/accountSieve the Sieve scripting page.

**Comment by ScottPeterJohnson** [comment]
Comment by **Scott** on May 21, 2019:
I appear to have made a pretty stupid typo on the example script.
```
if envelope "to" :user :matches "*" {
    set "user" "{1}";
}
if envelope "to" :domain :matches "*" {
    set "domain" "{1}";
}
```
This should actually be:
```
if envelope "to" :user :matches "*" {
    set "user" "${1}";
}
if envelope "to" :domain :matches "*" {
    set "domain" "${1}";
}
```
I.e. instead of "${1}" I had "{1}", which does not perform variable substitution, so the script attempts to redirect everything to {1}@{1}, which fails. Usually that failure would generate an error email to your primary address, but it appears I need to tweak the logic there slightly. 

The fallback of course routes normally.

So in summary:

> 1. Add that Sieve script by default to all accounts that don’t have Sieve scripts (so you don’t mess up what people already have) so that custom domains will support this by default to match your domains.

I do plan to do this, once I have a decent GUI for editing Sieve scripts. I guess I could do it beforehand though, I doubt anyone would mind.

> 2. Add a note on how long it takes for Sieve scripts to take effect on https://purelymail.com/manage/accountSieve the Sieve scripting page.

It takes effect immediately.

**Comment by ScottPeterJohnson** [comment]
Comment by **jeremy** on May 21, 2019:
Thanks for the quick follow-up.

I can't see why making a default Sieve script to cause custom domains to behave how your documentation states (yes, I know you have the caveat at the bottom) would offend anyone. 

I am unconvinced that a GUI implementation of Sieve is worthwhile, since power users can handle raw Sieve, and non-power users don't care. Maybe I don't know what I'm talking about. The fact that there is no open source GUI for Sieve available off the shelf already might be a strong signal, though.
